File: backend/app/models/extended_finance.py
import uuid

from sqlalchemy import Date, ForeignKey, Integer, Numeric, String, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import JSONB, TIMESTAMP, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database.session import Base


class CreditCardBill(Base):
    __tablename__ = "credit_card_bills"
    __table_args__ = (
        UniqueConstraint(
            "household_id", "pluggy_bill_id", name="uq_credit_card_bills_identity"
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    household_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("households.id"), nullable=False
    )
    account_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("accounts.id"), nullable=False
    )
    pluggy_bill_id: Mapped[str] = mapped_column(String, nullable=False)
    due_date: Mapped[Date | None] = mapped_column(Date, nullable=True)
    closing_date: Mapped[Date | None] = mapped_column(Date, nullable=True)
    total_amount: Mapped[float | None] = mapped_column(Numeric(18, 2), nullable=True)
    minimum_payment: Mapped[float | None] = mapped_column(Numeric(18, 2), nullable=True)
    currency_code: Mapped[str] = mapped_column(String(3), nullable=False, default="BRL")
    raw_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    created_at: Mapped[TIMESTAMP] = mapped_column(
        TIMESTAMP(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[TIMESTAMP] = mapped_column(
        TIMESTAMP(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )


class Investment(Base):
    __tablename__ = "investments"
    __table_args__ = (
        UniqueConstraint(
            "household_id", "pluggy_investment_id", name="uq_investments_identity"
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    household_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("households.id"), nullable=False
    )
    pluggy_connection_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("pluggy_connections.id"), nullable=False
    )
    pluggy_investment_id: Mapped[str] = mapped_column(String, nullable=False)
    name: Mapped[str | None] = mapped_column(String, nullable=True)
    type: Mapped[str | None] = mapped_column(String, nullable=True)
    subtype: Mapped[str | None] = mapped_column(String, nullable=True)
    balance: Mapped[float | None] = mapped_column(Numeric(18, 2), nullable=True)
    value: Mapped[float | None] = mapped_column(Numeric(18, 2), nullable=True)
    quantity: Mapped[float | None] = mapped_column(Numeric(18, 6), nullable=True)
    currency_code: Mapped[str] = mapped_column(String(3), nullable=False, default="BRL")
    investment_date: Mapped[Date | None] = mapped_column(Date, nullable=True)
    raw_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    created_at: Mapped[TIMESTAMP] = mapped_column(
        TIMESTAMP(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[TIMESTAMP] = mapped_column(
        TIMESTAMP(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )


class Loan(Base):
    __tablename__ = "loans"
    __table_args__ = (
        UniqueConstraint("household_id", "pluggy_loan_id", name="uq_loans_identity"),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    household_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("households.id"), nullable=False
    )
    pluggy_connection_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("pluggy_connections.id"), nullable=False
    )
    pluggy_loan_id: Mapped[str] = mapped_column(String, nullable=False)
    type: Mapped[str | None] = mapped_column(String, nullable=True)
    status: Mapped[str | None] = mapped_column(String, nullable=True)
    contract_amount: Mapped[float | None] = mapped_column(Numeric(18, 2), nullable=True)
    outstanding_balance: Mapped[float | None] = mapped_column(
        Numeric(18, 2), nullable=True
    )
    installment_amount: Mapped[float | None] = mapped_column(
        Numeric(18, 2), nullable=True
    )
    installments_total: Mapped[int | None] = mapped_column(Integer, nullable=True)
    installments_paid: Mapped[int | None] = mapped_column(Integer, nullable=True)
    due_date: Mapped[Date | None] = mapped_column(Date, nullable=True)
    interest_rate: Mapped[float | None] = mapped_column(Numeric(9, 4), nullable=True)
    currency_code: Mapped[str] = mapped_column(String(3), nullable=False, default="BRL")
    raw_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    created_at: Mapped[TIMESTAMP] = mapped_column(
        TIMESTAMP(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[TIMESTAMP] = mapped_column(
        TIMESTAMP(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )


class BalanceSnapshot(Base):
    __tablename__ = "balance_snapshots"
    __table_args__ = (
        UniqueConstraint(
            "account_id", "snapshot_date", name="uq_balance_snapshots_identity"
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    household_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("households.id"), nullable=False
    )
    account_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("accounts.id"), nullable=False
    )
    balance: Mapped[float | None] = mapped_column(Numeric(18, 2), nullable=True)
    currency_code: Mapped[str] = mapped_column(String(3), nullable=False, default="BRL")
    snapshot_date: Mapped[Date] = mapped_column(Date, nullable=False)
    created_at: Mapped[TIMESTAMP] = mapped_column(
        TIMESTAMP(timezone=True), nullable=False, server_default=func.now()
    )

File: frontend/lib/data/models/extended_finance.dart
class CreditCardBillSummary {
  const CreditCardBillSummary({
    required this.id,
    required this.accountId,
    required this.dueDate,
    required this.closingDate,
    required this.totalAmount,
    required this.minimumPayment,
    required this.currencyCode,
  });

  factory CreditCardBillSummary.fromJson(Map<String, dynamic> json) {
    return CreditCardBillSummary(
      id: json['id'] as String,
      accountId: json['account_id'] as String,
      dueDate: json['due_date'] == null
          ? null
          : DateTime.parse(json['due_date'] as String),
      closingDate: json['closing_date'] == null
          ? null
          : DateTime.parse(json['closing_date'] as String),
      totalAmount: (json['total_amount'] as num?)?.toDouble(),
      minimumPayment: (json['minimum_payment'] as num?)?.toDouble(),
      currencyCode: json['currency_code'] as String,
    );
  }

  final String id;
  final String accountId;
  final DateTime? dueDate;
  final DateTime? closingDate;
  final double? totalAmount;
  final double? minimumPayment;
  final String currencyCode;
}

class InvestmentSummary {
  const InvestmentSummary({
    required this.id,
    required this.name,
    required this.type,
    required this.subtype,
    required this.balance,
    required this.value,
    required this.quantity,
    required this.currencyCode,
    required this.investmentDate,
  });

  factory InvestmentSummary.fromJson(Map<String, dynamic> json) {
    return InvestmentSummary(
      id: json['id'] as String,
      name: json['name'] as String?,
      type: json['type'] as String?,
      subtype: json['subtype'] as String?,
      balance: (json['balance'] as num?)?.toDouble(),
      value: (json['value'] as num?)?.toDouble(),
      quantity: (json['quantity'] as num?)?.toDouble(),
      currencyCode: json['currency_code'] as String,
      investmentDate: json['investment_date'] == null
          ? null
          : DateTime.parse(json['investment_date'] as String),
    );
  }

  final String id;
  final String? name;
  final String? type;
  final String? subtype;
  final double? balance;
  final double? value;
  final double? quantity;
  final String currencyCode;
  final DateTime? investmentDate;
}

class LoanSummary {
  const LoanSummary({
    required this.id,
    required this.type,
    required this.status,
    required this.contractAmount,
    required this.outstandingBalance,
    required this.installmentAmount,
    required this.installmentsTotal,
    required this.installmentsPaid,
    required this.dueDate,
    required this.interestRate,
    required this.currencyCode,
  });

  factory LoanSummary.fromJson(Map<String, dynamic> json) {
    return LoanSummary(
      id: json['id'] as String,
      type: json['type'] as String?,
      status: json['status'] as String?,
      contractAmount: (json['contract_amount'] as num?)?.toDouble(),
      outstandingBalance: (json['outstanding_balance'] as num?)?.toDouble(),
      installmentAmount: (json['installment_amount'] as num?)?.toDouble(),
      installmentsTotal: json['installments_total'] as int?,
      installmentsPaid: json['installments_paid'] as int?,
      dueDate: json['due_date'] == null
          ? null
          : DateTime.parse(json['due_date'] as String),
      interestRate: (json['interest_rate'] as num?)?.toDouble(),
      currencyCode: json['currency_code'] as String,
    );
  }

  final String id;
  final String? type;
  final String? status;
  final double? contractAmount;
  final double? outstandingBalance;
  final double? installmentAmount;
  final int? installmentsTotal;
  final int? installmentsPaid;
  final DateTime? dueDate;
  final double? interestRate;
  final String currencyCode;
}

class BalancePoint {
  const BalancePoint({required this.snapshotDate, required this.totalBalance});

  factory BalancePoint.fromJson(Map<String, dynamic> json) {
    return BalancePoint(
      snapshotDate: DateTime.parse(json['snapshot_date'] as String),
      totalBalance: (json['total_balance'] as num).toDouble(),
    );
  }

  final DateTime snapshotDate;
  final double totalBalance;
}

class CategoryBreakdownItem {
  const CategoryBreakdownItem({
    required this.category,
    required this.total,
    this.previousTotal,
  });

  factory CategoryBreakdownItem.fromJson(Map<String, dynamic> json) {
    return CategoryBreakdownItem(
      category: json['category'] as String?,
      total: (json['total'] as num).toDouble(),
      previousTotal: (json['previous_total'] as num?)?.toDouble(),
    );
  }

  final String? category;
  final double total;
  final double? previousTotal;
}
